From d0ad66aacb633664a767aa6d5e45e9e3f6278920 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Mon, 22 Sep 2025 11:12:13 +0800
Subject: [PATCH] net: ethernet: mtk_eth_soc: add 2500Mbps maximum rate limt
 for NETSYSv3

Without this patch, the ETH driver is unable to set QDMA maximum rate
limit to 2500 Mbps for the specific TX queue.

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -848,6 +848,18 @@ static void mtk_set_queue_speed(struct m
 	      MTK_QTX_SCH_LEAKY_BUCKET_SIZE;
 	if (mtk_is_netsys_v1(eth))
 		val |= MTK_QTX_SCH_LEAKY_BUCKET_EN;
+		
+	
+	if (mtk_is_netsys_v3_or_greater(eth)) {
+		pr_info("mtk_eth: NETSYS version = 3\n");
+	} else if (mtk_is_netsys_v2_or_greater(eth)) {
+		pr_info("mtk_eth: NETSYS version = 2\n");
+	} else {
+		pr_info("mtk_eth: NETSYS version = 1\n");
+	}
+
+	if (!mtk_is_netsys_v3_or_greater(eth) && speed > SPEED_1000)
+		goto out;
 
 	if (IS_ENABLED(CONFIG_SOC_MT7621)) {
 		switch (speed) {
@@ -892,11 +904,17 @@ static void mtk_set_queue_speed(struct m
 			       FIELD_PREP(MTK_QTX_SCH_MAX_RATE_EXP, 6) |
 			       FIELD_PREP(MTK_QTX_SCH_MAX_RATE_WEIGHT, 10);
 			break;
+		case SPEED_2500:
+			val |= MTK_QTX_SCH_MAX_RATE_EN |
+			       FIELD_PREP(MTK_QTX_SCH_MAX_RATE_MAN, 25) |
+			       FIELD_PREP(MTK_QTX_SCH_MAX_RATE_EXP, 5) |
+			       FIELD_PREP(MTK_QTX_SCH_MAX_RATE_WEIGHT, 10);
+			break;
 		default:
 			break;
 		}
 	}
-
+out:
 	ofs = MTK_QTX_OFFSET * idx;
 	mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + ofs);
 }
